core/trino-main/src/test/java/io/trino/server/security/TestResourceSecurity.java

```diff
@@ -190,13 +202,66 @@ public void testPasswordAuthenticator()
                         .build())
                 .build()) {
             server.getInstance(Key.get(PasswordAuthenticatorManager.class)).setAuthenticator(TestResourceSecurity::authenticate);
-            server.getInstance(Key.get(AccessControlManager.class)).addSystemAccessControl(new TestSystemAccessControl());
+            server.getInstance(Key.get(AccessControlManager.class)).addSystemAccessControl(new TestSystemAccessControl(false));
             HttpServerInfo httpServerInfo = server.getInstance(Key.get(HttpServerInfo.class));
             assertAuthenticationDisabled(httpServerInfo.getHttpUri());
             assertPasswordAuthentication(httpServerInfo.getHttpsUri());
         }
     }
 
+    @Test
+    public void testPasswordAuthenticatorUserMapping()
+            throws Exception
+    {
+        try (TestingTrinoServer server = TestingTrinoServer.builder()
+                .setProperties(ImmutableMap.<String, String>builder()
+                        .putAll(SECURE_PROPERTIES)
+                        .put("http-server.authentication.type", "password")
+                        .put("http-server.authentication.password.user-mapping.pattern", ALLOWED_USER_MAPPING_PATTERN)
+                        .build())
+                .setAdditionalModule(binder -> jaxrsBinder(binder).bind(TestResource.class))
+                .build()) {
+            server.getInstance(Key.get(PasswordAuthenticatorManager.class)).setAuthenticator(TestResourceSecurity::authenticate);
+            server.getInstance(Key.get(AccessControlManager.class)).addSystemAccessControl(new TestSystemAccessControl(false));
+            HttpServerInfo httpServerInfo = server.getInstance(Key.get(HttpServerInfo.class));
+
+            // Test sets basic auth user and X-Trino-User, and the authenticator is performing user mapping.
+            // Normally this would result in an impersonation check to the X-Trino-User, but the password
+            // authenticator has a hack to clear X-Trino-User in this case.
+            Request request = new Request.Builder()
+                    .url(getLocation(httpServerInfo.getHttpsUri(), "/username"))
+                    .addHeader("Authorization", Credentials.basic(TEST_USER_LOGIN, TEST_PASSWORD))
+                    .addHeader("X-Trino-User", TEST_USER_LOGIN)
+                    .build();
+            try (Response response = client.newCall(request).execute()) {
+                assertEquals(response.code(), SC_OK);
+                assertEquals(response.header("user"), TEST_USER);
```

Comment:
Is this effective session user from the server?
Why would it be different from the one explicitly requested by the caller, with X-Trino-User header?

Reply:
> Is this effective session user from the server?
Yes!  See review comments above.

> Why would it be different from the one explicitly requested by the caller, with X-Trino-User header?
Because all of the methods in this test are using user mapping.  If you look above you will see that `X-Trino-User` is set to `TEST_USER_LOGIN` and that is `test-user@allowed`,  the configuration mapps that to `test-user` which is the value of the constant `TEST_USER` I am using here.